content/en/blog/general/why-chainsaw-is-unique/index.md

```diff
@@ -0,0 +1,200 @@
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  labels:
+    k8s-app: kube-dns
+  namespace: kube-system
+spec:
+  template:
+    spec:
+      (containers[?securityContext == null]):
+        (len(@)): 0
+```
+
+In the assertion above, the first three fields `spec`, `template`, and `spec` are basic projections that simply take the content of their respective fields and pass that to descendants.
+
+`(containers[?securityContext == null])` is a JMESPath expression filtering the `containers` array, selecting only the element where `securityContext` is `null`. This projection results in a new array that is passed to the descendant (`(len(@))` in this case).
+
+`(len(@))` is another JMESPath expression that computes the length of the array. There's no more descendant at this point. We're at a leaf of the YAML tree and the array length we just computed is then compared against 0.
+
+If the comparison matches, the assertion will be considered valid; if not, it will be considered failed.
+
+In simple words, the assertion above checks that all containers in the Deployment have a security context configured.
+
+### Iterating over arrays (or maps)
+
+Being able to filter arrays allows selecting the elements to be processed.
+
+On top of that, Chainsaw allows iterating over array elements to validate each item separately.
+
+The assertion above can be rewritten using an iterator:
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  labels:
+    k8s-app: kube-dns
+  namespace: kube-system
+spec:
+  template:
+    spec:
+      # the ~ modifier tells Chainsaw to iterate over the array elements
+      ~.(containers):
+        securityContext: {}
+```
+
+This assertion uses the `~` modifier and Chainsaw will evaluate descendants once per element in the array.
+
+### Explicit bindings and reference
+
+When passing from a parent node to descendant nodes, it can be useful to keep a reference to one or more parent nodes.
+
+Let's say we want to consider the pod-level security context in addition to the container-level security context. We can write such an assertion with the syntax below:
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  labels:
+    k8s-app: kube-dns
+  namespace: kube-system
+spec:
+  template:
+    (spec)->specBinding:
+      # the ~ modifier tells Chainsaw to iterate over the array elements
+      ~.(containers):
+        ($specBinding.securityContext != null || securityContext != null): true
+```
+
+In the assertion above, `(spec)->specBinding` means that Chainsaw will keep a reference to the result of the expression evaluation and will make it available to descendants under the `$specBinding` binding.
+
+Now we can use the binding to write the assertion check taking the pod level security context into account like this `($specBinding.securityContext != null || securityContext != null): true`.
+
+### Give me more
```

Comment:
I don't like this title, @chipzoller do you have a suggestion for a better one ?

Reply:
(above is a run-on sentence ? ;-) )

Reply:
I don't like this title. @chipzoller do you have a suggestion for a better one ?

Reply:
## What else

Reply:
"More" is fine.

Reply:
Like What more ?